Algorithm.md

```diff
@@ -0,0 +1,30 @@
+1. Fetch available versions for each dependency
+2. If a project is already in the graph, intersect the new requirement with the one already present
+3. Select the highest [precedence] version which satisfies the intersected requirement (or only requirement, if this is the first insertion)
+  * This always has to be O(n) over the list of versions, because custom requirements mean that a previously failing version might become a candidate again
+4. Collect all child dependencies from that graph depth, repeat from step 1
+
+### In event of a conflict
+
+Given a project already in the graph with requirement X, and about to be re-added to the graph with requirement Y, where X and Y conflict:
+
+1. Find the _dependent_ project which changed the requirement into X (either from nil, or from a previously looser requirement)
+2. Select the next best version of that dependent project and attempt to finish resolving
+3. In the event that no other versions satisfy the requirements applied to the dependent, we have two choices:
+  - GOTO step 1 for the dependent project, or
+  - Find a further back time where the originally conflicting dependency (the one with requirement X) was added (e.g., with requirement W), and try replacing the dependent which added it at _that_ level instead
```

Comment:
Suggesting implementation considerations. You can, to some extent, pre-order the dependencies to achieve the above effect without dynamically re-ordering the search.

Here's how it would work, roughly:

1. Take your root dependency list and examine their dependencies at the most recent compatible version
2. Order all dependencies in build order and select just the root projects. This prevents backtracking as pruning occurs.
3. Then search through all remaining dependencies in build order.

Since constraints mostly affect unvisited nodes, backtracking is reduced.

Other optimizations:

1. When selecting an earlier version of a project, only choose one with different version constraints on its dependencies
2. The book also describes a _backmarking_ algorithm that may be useful